src/Conversion/ONNXToTOSA/Tensor/Resize.cpp

```diff
@@ -0,0 +1,314 @@
+    return tosa::convertNegativeAxis(attr.getInt(), 4);
+  });
+}
+
+class ONNXResizeOpLoweringToTOSA : public ConversionPattern {
+public:
+  ONNXResizeOpLoweringToTOSA(MLIRContext *ctx)
+      : ConversionPattern(ONNXResizeOp::getOperationName(), 1, ctx) {}
+  using OpAdaptor = typename ONNXResizeOp::Adaptor;
+
+  struct ScaleHelper {
+    ScaleHelper(
+        int64_t numerator, int64_t denominator, int64_t offset, int64_t border)
+        : numerator(numerator), denominator(denominator), offset(offset),
+          border(border){};
+    int64_t numerator, denominator, offset, border;
+  };
+
+  struct FractionNumber {
+    FractionNumber(int64_t numerator, int64_t denominator)
+        : numerator(numerator), denominator(denominator){};
+    FractionNumber(float number) {
+      double integral = std::floor(number);
+      double frac = number - integral;
+
+      const long precision = 1000000000; // This is the accuracy.
+
+      long gcd_ = std::gcd((int)round(frac * precision), precision);
+
+      long denominator = precision / gcd_;
+      long numerator = round(frac * precision) / gcd_;
+
+      this->numerator = numerator + denominator * integral;
+      this->denominator = denominator;
+    }
+    int64_t numerator;
+    int64_t denominator;
+  };
+
+  /// ## coordinateTransformationMode ##
+  /// TOSA uses the formula ix = (ox * scale_x_d + offset_x) / scale_x_n
+  /// to find the input coordinates. In order to lower ONNX one needs to
+  /// express the modes in this context. Border is only used to check if certain
+  /// conditions in the dimensions are met, but has no actual use in calculating
+  /// something. It is probably an error in the TOSA specification. In order to
+  /// fulfill the conditions, border is always
+  /// border = d * (output - 1) - n *(input - 1) + offset
+  ///
+  /// ### half_pixel ###
+  /// ONNX formula: ix = (ox + 0.5) / scale - 0.5
+  /// gcd = greatest common divisor
+  /// To meet TOSA requirements:
+  /// - scale_x_d = input_size * 2 / gcd
+  /// - scale_x_n = output_size * 2 / gcd
+  /// - offset_x =  (scale_x_d - scale_x_n) / 2
+  ///
+  /// ### pytorch_half_pixel ###
+  /// Same as half_pixel, but if output == 1:
+  /// - scale_x_d = scale_x_n = 1
+  /// - offset = -1
+  ///
+  /// ### half_pixel_symmetric ###
+  /// NOT SUPPORTED BY TOSA, BECAUSE OF INT OFFSET
+  /// Same as half_pixel, but with another offset
+  /// - offset_x += symmetric_offset_x
+  ///
+  /// ### align_corners ###
+  /// - scale_x_d = (input_size - 1)
+  /// - scale_x_n = (output_size - 1)
+  /// - offset = 0
+  ///
+  /// ### asymmetric ###
+  /// - scale_x_d = input_size
+  /// - scale_x_n = output_size
+  /// - offset = 0
+  ///
+  /// ## nearest_mode ##
+  /// If mode == nearest, then ONNX can set the nearest_mode attribute.
+  /// The standard case for TOSA of round_half_up. Support for floor can
+  /// be achieved with:
+  /// offset_x -= n/2
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
```

Comment:
Another suggestion, could we break up this matchAndRewrite into small parts. I find it hard to follow at times. I think it may help to break it up into pieces, i.e. smaller functions. Because at the moment its a 200 line function. 

What do you think?

Reply:
I will move the lambda and the scale value handling into the own functions. The lambda will get a lot more arguments, but I guess that is still better than now